### PR TITLE
Add chainstate-test-framework::PoSBlockBuilder

### PR DESCRIPTION
**Pull request by azarovh**
Introduce `PoSBlockBuilder` that should be a convenient tool to create PoS blocks in unit tests. I reworked one test in `pos_processing_tests.rs` as an example, more tests can be reworked later. So I think #972 can be closed.

Though the main motivation for this change was that there were no way to test long chain reorgs for PoS. In #967 I introduced in-memory reorg which was effectively not tested after I disabled it for PoW.

### CODE REVIEW

**Review by TheQuantumPhysicist — APPROVED** [review]
Besides the minor comments I made, looking good.

**Review by muursh — APPROVED** [review]
I'm fine with this

File: chainstate/test-suite/benches/benches.rs [hunk 1]
```diff
@@ -14,24 +14,88 @@
...
 
-criterion_group!(benches, reorg);
+pub fn pos_reorg(c: &mut Criterion) {
+    let mut rng = make_seedable_rng(1111.into());
+    let (staking_sk, staking_pk) = PrivateKey::new_from_rng(&mut rng, KeyKind::Secp256k1Schnorr);
+    let (vrf_sk, vrf_pk) = VRFPrivateKey::new_from_rng(&mut rng, VRFKeyKind::Schnorrkel);
+
+    let pool_id = PoolId::new(H256::random_using(&mut rng));
+    let stake_pool_data = StakePoolData::new(
+        Amount::from_atoms(40_000_000 * Mlt::ATOMS_PER_MLT),
```
Comment (alfiedotwtf) [inline chainstate/test-suite/benches/benches.rs:58]: Should this use `chain_config.min_stake_pool_pledge()`?

File: chainstate/test-suite/src/tests/pos_accounting_reorg.rs [hunk 2]
```diff
@@ -232,3 +233,58 @@ fn stake_pool_reorg(#[case] seed: Seed) {
...
+fn long_chain_reorg(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let (staking_sk, staking_pk) =
+            PrivateKey::new_from_rng(&mut rng, KeyKind::Secp256k1Schnorr);
+        let (vrf_sk, vrf_pk) = VRFPrivateKey::new_from_rng(&mut rng, VRFKeyKind::Schnorrkel);
+
+        let pool_id = PoolId::new(H256::random_using(&mut rng));
+        let stake_pool_data = StakePoolData::new(
+            Amount::from_atoms(40_000_000 * Mlt::ATOMS_PER_MLT),
```
Comment (alfiedotwtf) [inline chainstate/test-suite/src/tests/pos_accounting_reorg.rs:249]: Maybe `chain_config.min_stake_pool_pledge()` here too.

File: chainstate/test-suite/src/tests/pos_processing_tests.rs [hunk 3]
```diff
@@ -2103,30 +2101,9 @@ fn spend_from_delegation_with_reward(#[case] seed: Seed) {
...
-    let chain_config = ConfigBuilder::test_chain().net_upgrades(net_upgrades).build();
-
-    let mut tf = TestFramework::builder(&mut rng).with_chain_config(chain_config).build();
 
-    // Process block_1 and create stake pool and delegation
-    let min_pledge_atoms = tf.chainstate.get_chain_config().min_stake_pool_pledge().into_atoms();
-    let amount_to_stake = Amount::from_atoms(rng.gen_range(min_pledge_atoms..min_pledge_atoms * 3));
-    let amount_to_delegate = Amount::from_atoms(rng.gen_range(100..100_000));
+    let pool_id = PoolId::new(H256::random_using(&mut rng));
+    let amount_to_stake = Amount::from_atoms(40_000_000 * Mlt::ATOMS_PER_MLT);
```
Comment (alfiedotwtf) [inline chainstate/test-suite/src/tests/pos_processing_tests.rs:2106]: `chain_config.min_stake_pool_pledge()` here too.

**Review by alfiedotwtf — APPROVED** [review]

**Comment by azarovh** [comment]
> Besides the minor comments I made, looking good.

@TheQuantumPhysicist I don't see any

File: chainstate/test-suite/benches/benches.rs (hunk 1 above)
Comment (azarovh) [reply chainstate/test-suite/benches/benches.rs:58]: the problem is that ChainConfig is not constructed yet. The same with other places where I put the pool into genesis block. But I agree that it doesn't look good, will try to find a place for a constant

Comment (alfiedotwtf) [reply chainstate/test-suite/benches/benches.rs:58]: Yeah, fair enough. I created a throw away in `blockprod`. Maybe wasteful, but long term better maintainable:

```
let min_stake_pool_pledge = {                    
    // throw away just to get value              
    let chain_config = create_unit_test_config();
    chain_config.min_stake_pool_pledge()         
};                                               
```

**Comment by TheQuantumPhysicist** [comment]
> > Besides the minor comments I made, looking good.
> 
> @TheQuantumPhysicist I don't see any

This is probably vscode not clearing up the text field after approval of another commit with this message. Never mind.

**Review by iljakuklic — APPROVED** [review]

File: chainstate/test-suite/benches/benches.rs (hunk 1 above)
Comment (azarovh) [reply chainstate/test-suite/benches/benches.rs]: used `create_unit_test_config ` as well

**Pull request merged**